src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs

```diff
@@ -107,9 +133,43 @@ private class GlobalImportAddRemoveTagHelperVisitor : TagHelperDirectiveSpanVisi
                 IEnumerable<TagHelperDirectiveDescriptor> descriptors,
                 ParserErrorSink errorSink)
             {
-                return base.GetTagHelperDescriptorResolutionContext(
-                    _globalImportDirectiveDescriptors.Concat(descriptors),
-                    errorSink);
+                var directivesToImport = MergeDirectiveDescriptors(descriptors, _globalImportDirectiveDescriptors);
+
+                return base.GetTagHelperDescriptorResolutionContext(directivesToImport, errorSink);
+            }
+
+            private static IEnumerable<TagHelperDirectiveDescriptor> MergeDirectiveDescriptors(
+                IEnumerable<TagHelperDirectiveDescriptor> descriptors,
+                IEnumerable<TagHelperDirectiveDescriptor> inheritedDescriptors)
+            {
+                var mergedDescriptors = new List<TagHelperDirectiveDescriptor>();
+                TagHelperDirectiveDescriptor prefixDirectiveDescriptor = null;
+
+                foreach (var descriptor in inheritedDescriptors)
+                {
+                    if (descriptor.DirectiveType != TagHelperDirectiveType.TagHelperPrefix)
```

Comment:
Super nit, maybe change the conditionals so setting the `prefixDirectiveDescriptor` is closer to checking if it's a `TagHelperPrefix` directive

```C#
if (descriptor.DirectiveType == TagHelperDirectiveType.TagHelperPrefix)
{
     prefixDirectiveDescriptor = descriptor;
}
else
{
   mergedDescriptors.Add(descriptor);
}
```